src/block_ffm.rs

```diff
@@ -542,16 +554,53 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
                             is_first_feature = false;
                             contra_fields_copied = true;
                             // Copy only once, skip other copying as the data for all features of that contra_index is already calculated
-                            contra_fields
-                                .get_unchecked_mut(offset..offset + field_embedding_len_as_usize)
-                                .copy_from_slice(
-                                    cached_contra_fields.get_unchecked(
-                                        offset..offset + field_embedding_len_as_usize,
-                                    ),
-                                );
+                            ptr::copy_nonoverlapping(
+                                cached_contra_fields.as_ptr().add(offset),
+                                contra_fields.as_mut_ptr().add(offset),
+                                field_embedding_len_as_usize,
+                            );
                         } else if !contra_fields_copied {
                             contra_fields_copied = true;
-                            for z in 0..field_embedding_len_as_usize {
+
+                            const LANES: usize = STEP * 4;
+                            let field_embedding_len_end = field_embedding_len_as_usize
+                                - (field_embedding_len_as_usize % LANES);
+
+                            let mut contra_fields_ptr = contra_fields.as_mut_ptr().add(offset);
+                            let mut cached_contra_fields_ptr =
+                                cached_contra_fields.as_ptr().add(offset);
+
+                            for _ in (0..field_embedding_len_end).step_by(LANES) {
+                                add_cached_contra_field(
+                                    contra_fields_ptr,
+                                    cached_contra_fields_ptr,
+                                );
+                                contra_fields_ptr = contra_fields_ptr.add(STEP);
+                                cached_contra_fields_ptr = cached_contra_fields_ptr.add(STEP);
```

Comment:
What's the expected memory overhead of this cache btw?

Reply:
There is no real overhead as its done per batch. & normally its a set of values for the number of cached features. So very small

Reply:
Right, this seems negligible w.r.t. weight space then